domain/agentbinary/service/controllerstore.go

```diff
@@ -0,0 +1,205 @@
+	// the object store but unless the association has been made this will
+	// always return false.
+	CheckAgentBinarySHA256Exists(context.Context, string) (bool, error)
+
+	// GetObjectUUID returns the object store UUID for the given object path.
+	// The following errors can be returned:
+	// - [agentbinaryerrors.ObjectNotFound] when no object exists that matches this path.
+	GetObjectUUID(ctx context.Context, path string) (objectstore.UUID, error)
+
+	// RegisterAgentBinary registers a new agent binary's metadata to the database.
+	// [agentbinaryerrors.AlreadyExists] when the provided agent binary already
+	// exists.
+	// [agentbinaryerrors.ObjectNotFound] when no object exists that matches
+	// this agent binary.
+	// [coreerrors.NotSupported] if the architecture is not supported by the
+	// state layer.
+	RegisterAgentBinary(ctx context.Context, arg agentbinary.RegisterAgentBinaryArg) error
+
+	// GetAgentBinarySHA256 checks if an agent binary record for the given
+	// version exists and, if found, returns its associated SHA256
+	// value. Note that this only verifies existence of a recorded agent binary entry,
+	// the underlying object may exist in the store without this association, in which case
+	// the function will return false.
+	GetAgentBinarySHA256(ctx context.Context, ver coreagentbinary.Version, stream coreagentbinary.AgentStream) (bool, string, error)
+}
+
+type ControllerAgentBinaryStore struct {
+	state             ControllerStoreState
+	logger            logger.Logger
+	objectStoreGetter objectstore.NamespacedObjectStoreGetter
+}
+
+// NewControllerAgentBinaryStore returns a new instance of ControllerAgentBinaryStore.
+func NewControllerAgentBinaryStore(
+	state ControllerStoreState,
+	logger logger.Logger,
+	objectStoreGetter objectstore.NamespacedObjectStoreGetter,
+) *ControllerAgentBinaryStore {
+	return &ControllerAgentBinaryStore{
+		state:             state,
+		logger:            logger,
+		objectStoreGetter: objectStoreGetter,
+	}
+}
+
+// GetAgentBinary retrieves the agent binary corresponding to the given version
+// and stream from the controller's object store.
+//
+// The function first queries controller state to check whether a SHA256 record
+// exists for the requested version and stream. If no such record is found,
+// agentbinaryerrors.NotFound is returned.
+//
+// If a valid SHA256 is recorded, the corresponding binary blob is fetched from
+// the controller's object store and returned as an io.ReadCloser together with
+// its size and SHA256 string.
+//
+// The caller is responsible for closing the returned reader.
+func (s *ControllerAgentBinaryStore) GetAgentBinary(ctx context.Context, ver coreagentbinary.Version, stream coreagentbinary.AgentStream) (io.ReadCloser, int64, string, error) {
+	hasAgentBinary, sha256Str, err := s.state.GetAgentBinarySHA256(ctx, ver, stream)
+	if err != nil {
+		return nil, 0, "", err
+	}
+	if !hasAgentBinary {
+		return nil, 0, "", agentbinaryerrors.NotFound
```

Comment:
Happy for this to be in a follow up PR:

```suggestion
		return nil, 0, "", errors.New("agent binary not found in controller store").Add(agentbinaryerrors.NotFound)
```